### DISCUSSION TITLE
cfengine-community Debian repository is defective

### DISCUSSION THREAD

**Discussion by fluchtkapsel** [cfengine/core#5506] [General]
apt selects the wrong package on Debian 11:
```
root@gamma:~# apt install cfengine-community 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  cfengine-community
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 5,139 kB of archives.
After this operation, 19.2 MB of additional disk space will be used.
Get:1 https://cfengine-package-repos.s3.amazonaws.com/pub/apt/packages stable/main amd64 cfengine-community amd64 3.23.0-1.debian9 [5,139 kB]
Fetched 5,139 kB in 2s (3,177 kB/s)           
Selecting previously unselected package cfengine-community.
(Reading database ... 67758 files and directories currently installed.)
Preparing to unpack .../cfengine-community_3.23.0-1.debian9_amd64.deb ...
WARNING: this package was built on debian 9, and you're installing it on debian 11, which seems to be newer.
This configuration might be unsupported, please consider downloading a proper package.
Press Ctrl+C within next 15 seconds to abort.
Unpacking cfengine-community (3.23.0-1.debian9) ...
Setting up cfengine-community (3.23.0-1.debian9) ...
```

After removing the package and installing the downloaded .deb file from the downloads page:
```
root@gamma:~# wget https://cfengine-package-repos.s3.amazonaws.com/community_binaries/Community-3.23.0/agent_debian11_x86_64/cfengine-community_3.23.0-1.debian11_amd64.deb
--2024-05-10 14:32:07--  https://cfengine-package-repos.s3.amazonaws.com/community_binaries/Community-3.23.0/agent_debian11_x86_64/cfengine-community_3.23.0-1.debian11_amd64.deb
Resolving cfengine-package-repos.s3.amazonaws.com (cfengine-package-repos.s3.amazonaws.com)... 52.216.110.91, 52.217.197.57, 3.5.3.139, ...
Connecting to cfengine-package-repos.s3.amazonaws.com (cfengine-package-repos.s3.amazonaws.com)|52.216.110.91|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5372052 (5.1M) [binary/octet-stream]
Saving to: ‘cfengine-community_3.23.0-1.debian11_amd64.deb’

cfengine-community_3.23.0-1.debian11_amd64.deb        100%[======================================================================================================================>]   5.12M  5.13MB/s    in 1.0s    

2024-05-10 14:32:09 (5.13 MB/s) - ‘cfengine-community_3.23.0-1.debian11_amd64.deb’ saved [5372052/5372052]

root@gamma:~# dpkg -i cfengine-community_3.23.0-1.debian11_amd64.deb 
(Reading database ... 68402 files and directories currently installed.)
Preparing to unpack cfengine-community_3.23.0-1.debian11_amd64.deb ...
Unpacking cfengine-community (3.23.0-1.debian11) over (3.18.7-1.debian11) ...
Setting up cfengine-community (3.23.0-1.debian11) ...
```

So, the package for Debian 11 is clearly there. Some further experimentation:
```
root@gamma:~# apt-cache policy cfengine-community 
cfengine-community:
  Installed: 3.23.0-1.debian11
  Candidate: 3.23.0-1.debian11
  Version table:
 *** 3.23.0-1.debian11 100
        100 /var/lib/dpkg/status
     3.23.0-1.debian9 500
        500 https://cfengine-package-repos.s3.amazonaws.com/pub/apt/packages stable/main amd64 Packages
     3.22.0-1.debian9 500
[…]
root@gamma:~# apt purge cfengine-community 
[…]
Removing cfengine-community (3.23.0-1.debian11) ...
[…]
root@gamma:~# cat /etc/apt/sources.list.d/cfengine-community.list ~^C
root@gamma:~# apt-cache policy cfengine-community 
cfengine-community:
  Installed: (none)
  Candidate: 3.23.0-1.debian9
  Version table:
     3.23.0-1.debian9 500
        500 https://cfengine-package-repos.s3.amazonaws.com/pub/apt/packages stable/main amd64 Packages
     3.22.0-1.debian9 500
        500 https://cfengine-package-repos.s3.amazonaws.com/pub/apt/packages stable/main amd64 Packages
[…]
```

Checking the list file:
```
root@gamma:~# grep -e "^Version:" /var/lib/apt/lists/cfengine-package-repos.s3.amazonaws.com_pub_apt_packages_dists_stable_main_binary-amd64_Packages 
Version: 3.4.5-1
Version: 3.12.1-1
Version: 3.9.1-1
Version: 3.18.0-1.debian7
Version: 3.12.5-1.debian7
Version: 3.8.2-1
Version: 3.12.3-2
Version: 3.13.0-1
Version: 3.12.3-1
Version: 3.8.1-1
Version: 3.12.4-1.7
Version: 3.11.0-1
Version: 3.7.0-1
Version: 3.21.4-1.debian9
Version: 3.22.0-1.debian9
Version: 3.14.0-1
Version: 3.7.6-1
Version: 3.12.0-1
Version: 3.6.4-1
Version: 3.16.0-1.debian7
Version: 3.6.5-1
Version: 3.17.0-1.debian7
Version: 3.15.2-1.debian7
Version: 3.10.0-1
Version: 3.12.7-1.debian7
Version: 3.15.5-1.debian9
Version: 3.19.0-1.debian9
Version: 3.21.1-1.debian9
Version: 3.18.2-1.debian9
Version: 3.15.0-2
Version: 3.9.0-1
Version: 3.15.3-1.debian7
Version: 3.23.0-1.debian9
Version: 3.18.3-2.debian9
Version: 3.9.2-1
Version: 3.18.1-1.debian9
Version: 3.20.0-1.debian9
Version: 3.5.1-3
Version: 3.7.4-1
Version: 3.7.5-1
Version: 3.4.4-1
Version: 3.7.8-1
Version: 3.8.0-1
Version: 3.6.2-1
Version: 3.10.2-1
Version: 3.10.6-1
Version: 3.5.3-1
Version: 3.6.7-1
Version: 3.21.0-1.debian9
Version: 3.12.6-1.debian7
Version: 3.21.3-1.debian9
Version: 3.6.3-1
Version: 3.15.1-1.debian7
Version: 3.10.1-1
Version: 3.7.3-1
Version: 3.18.6-1.debian9
Version: 3.10.5-1
Version: 3.15.6-1.debian9
Version: 3.18.3-1.debian9
Version: 3.7.7-1
Version: 3.12.4-2.7
Version: 3.7.1-1
Version: 3.10.3-1
Version: 3.7.2-1
Version: 3.21.2-1.debian9
Version: 3.6.0-1
Version: 3.10.4-1
Version: 3.12.2-2
Version: 3.18.7-1.debian9
Version: 3.15.7-1.debian9
Version: 3.10.7-1
Version: 3.6.6-1
Version: 3.18.5-1.debian9
Version: 3.15.0-1
Version: 3.6.1-1
Version: 3.5.2-1
Version: 3.18.4-1.debian9
Version: 3.15.4-1.debian7
```

There no packages for Debian 10, 11 or 12. I'd say whatever generates those files hasn't been updated to reflect packages for those Debian versions.

Another thing I noticed is that installing the LTS version 3.18.7 from the .deb file still leads to apt suggesting an upgrade to the non-LTS version 3.23.0:
```
root@gamma:~# apt list --upgradable 
Listing... Done
cfengine-community/stable 3.23.0-1.debian9 amd64 [upgradable from: 3.18.7-1.debian11]
```

**Comment by craigcomstock** [comment]
You are right in observing that we only provide packages marked as for debian 9 in our repository. This is the same as for the yum repository. Generally speaking these packages marked for older distribution version will work but may encounter trouble sometimes with system dependencies.

Did you try the debian-9 package on debian-11 and see if it worked or not?

---

Docs are here for our repositories: https://cfengine.com/cfengine-linux-distros/

There are three alternatives to using the repos: cf-remote, quick install script and download pages. These are the three options we typically use and would recommend.

```
pip install cf-remote
cf-remote install --clients root@host # install latest LTS
cf-remote --version master install --clients root@host # latest nightly
```

The quick install script uses the packages available on the download page and selects the most recent LTS by default:

https://cfengine.com/downloads/quick-install/

Here are the download pages for https://cfengine.com/downloads/cfengine-community/ and enterprise https://cfengine.com/downloads/cfengine-enterprise/

---

As for the upgrade path that's a bit unavoidable as many distributions will package our latest release regardless of LTS status and so then only the version number comparison is in effect.

Hope that clears things up a bit. Let us know how we can help.

**Reply by fluchtkapsel** [reply]
No, I didn't try the Debian 9 package on Debian 11. I was just wondering why the repository doesn't index the Debian 11 packages as they are clearly there, too.

One of the reasons we chose CFEngine after having some "fun" with Puppet and Saltstack was the fact that installing CFEngine didn't require any additional dependencies. Installing Python and pip and a pip package seems to counteract this. Set up a machine, install CFEngine, connect with server, done.

Also, already providing a repository and then pointing to a Python package … I'm sure, there's a story why it's the way it is. And I'd like to understand it.

My worst assumption at the moment is that marketing the CFEngine Enterprise edition takes precedence over providing a CFEngine Community repository which might lead people to just use that instead of paying for Enterprise. That'd be nothing I'd welcome, but something I could understand.

Another assumption is that using AWS causes just too much costs for repositories which provide regular updates to thousands or tens of thousands of clients.

Maybe I'll try the quickinstall script. I'll take a look.

**Reply by craigcomstock** [reply]
The quickinstall script only works with enterprise packages unfortunately.

The reason Debian 11 packages aren't in the repository is only for lack of time and resources. We have an internal ticket waiting for attention in this regard. If a community member was interested in putting in some effort that would be appreciated.

We wrote cf-remote in Python because it is a developer tool and so benefits from the higher level language available in Python. We primarily use it from one host via ssh credentials but other parts of CFEngine can benefit from Python as well such as custom promise modules and packages modules.

You can easily install community with cf-remote with an optional argument:

```
cf-remote install --edition community --clients host1,host2,host3
```

Which will by default install the LTS, currently 3.21.4.

Most of our marketing efforts are done by developers like me and are focused on showing folks generally how CFEngine works, not specifically enterprise or community. :)

Enterprise is free for 25 hosts so hobbyists and learners can easily get started that way.

Hopefully we can get our repositories fixed up with the latest versions and for currently supported versions of distributions soon.

**Reply by fluchtkapsel** [reply]
@craigcomstock Thanks for the answer, that helps a lot. :-)

As tempting as it might be to try to put the effort in you mentioned I don't know if I'm up to the task. When asking on what'd required the answers might take so much time that doing it yourself might be quicker :laughing:.

**Reply by olehermanse** [reply]
@MasinAD Everything Craig wrote is correct, but I can perhaps add a bit more context;

> Also, already providing a repository and then pointing to a Python package … I'm sure, there's a story why it's the way it is. And I'd like to understand it.

I wrote the initial version of `cf-remote` years ago, so I can share a bit of the history;

We've always provided packages for download on our website. (rpms, debs, msis, etc.). Downloading and installing them is still easy, and relatively common. Or putting them in your own repo server is another common approach. I'd consider all of this quite normal / standard ways of installing CFEngine among our users.

On top of this, we have a quick install script. It basically allows you to copy a bash one-liner, and it will find and install the appropriate package for your host. From a user's perspective, it works well, at least for one or two hosts, but it doesn't really help you if you want something more (older version, more hosts, community instead of enterprise, etc.). From our perspective, it's always been kind of a maintenance burden, we have to update this script for every release, fix issues with platform detection, etc. It's not something we really want to expand, but we do see that our users find it valuable, so we've kept it around.

Years ago, whenever I was manually testing new versions of CFEngine or even just an arbitrary package with my new features in it, my workflow could look like this:

1. Go to AWS or DigitalOcean to spawn some VMs. Maybe I want to test a combination of Ubuntu and RHEL, or maybe even Windows. I would also use local VMs some of the times (with VirtualBox and Vagrant).
2. Note down the IP addresses of the machines I've spawned.
3. Also add the appropriate SSH username which should be used for each of them (sometimes it's `root`, sometimes it's `ubuntu`, `vagrant`, `ec2-user`, `Administrator`, etc.
4. Go to the website / the appropriate CI pipeline to find the .deb, .rpm, .msi packages I want to install.
5. Run `scp` commands to transfer each package to the appropriate host.
6. Use `ssh` to log in to each host (On Windows I'd often have to first start powershell to get a more sane shell than `cmd.exe`).
7. Run `dpkg`, `rpm`, or `misexec` to install the appropriate package.
8. Run `cf-agent --bootstrap` with the appropriate IP to actually start CFEngine.
9. Actually start testing something (scp over some policy, look at logs, open the Web UI, or something else).

If you think that sounds tedious, we agree. I started putting all of those things into a script which eventually became `cf-remote`. I chose Python because it allowed me to implement things effectively, and it only needed to run on one machine, having some python on that 1 point, was not a problem. Thanks to several contributions from other team members, all of these steps have been implemented and automated, and the resulting experience of using `cf-remote` is quite nice, IMO. If you'd like to see an example, see this script:

https://github.com/cfengine/core/blob/master/docs/cf_remote_demo.sh

I use that script relatively often to spawn 6 different machines and install CFEngine nightly packages. It's also really easy to update if I want to also deploy some policy, or install a different version (like 3.18, or community).

Over time, we've added more nice functionality to cf-remote. If you're interested in what you can do with it, see:

https://cfengine.com/blog/2024/feature-friday-4-cf-remote/

https://pypi.org/project/cf-remote/

So, that leaves us with 4 ways of installing CFEngine:

1. Packages directly from download pages
2. Quick install bash script
3. cf-remote python tool
4. Apt / yum package repos

Each of these has its advantages and disadvantages. If you don't want to use python / python tools, then `cf-remote` is not a good option, but I'd just like to point out that **the requirement for python is only for the one machine running cf-remote**. Typically, this is your own workstation / laptop, or maybe the CFEngine hub, if that's preferable.

**`cf-remote` does not require or run any python on the individual hosts where you install CFEngine.**

Also, keep in mind what platforms and installers we currently support;

1. A combination of different architectures - 32 bit, 64 bit, x86, ARM, Itanium, PowerPC, SPARC.
2. OSes - Ubuntu, Debian, RHEL, CentOS, Windows, SUSE, Solaris, HP-UX, IBM AIX. (Also, several people do build and run CFEngine on various BSDs and macOS).
3. `.deb`, `.rpm`, `.msi`, `.bff`, `.tar.gz`, and a "classic" `./configure; make; make install` approach.

So, all of this takes time to maintain and support, and helps explain the desire for a tool like `cf-remote`, which is designed to work the same regardless of where you're installing CFEngine, and thus can help abstract away some of this complexity for the users. (It does not support all the platforms above yet, but many of them). It also relies on the same JSON information as our website, and which our CI produces, meaning it is quite good at handling new OS versions, new versions of CFEngine, nightly builds, etc.

> My worst assumption at the moment is that marketing the CFEngine Enterprise edition takes precedence over providing a CFEngine Community repository which might lead people to just use that instead of paying for Enterprise. That'd be nothing I'd welcome, but something I could understand.

As Craig said, it's just a matter of time and resources. There is no separate team for Enterprise or anything like that. It's one team working on both CFEngine Community and Enterprise, and as Craig pointed out, we are here communicating with you, and you can go to our blog and see blog posts from the same people, etc.

Paying customers can request features or platforms to be supported, or pay extra for us to provide support for something, and we do naturally prioritize that, but that's really it, there's no hidden agenda.

> No, I didn't try the Debian 9 package on Debian 11. I was just wondering why the repository doesn't index the Debian 11 packages as they are clearly there, too.

As Craig said, we have the oldest supported version there in the hope that it will also work on newer versions. If you, or anyone else, would like to help by maintaining some of these package repos (apt, yum, brew, arch, etc.) we'd applaud it, and be happy to collaborate with you.

Hopefully that longer explanation and history makes sense, let me know if you have any other questions :)

**Reply by fluchtkapsel** [reply]
@olehermanse Wow! Thanks for this exhausting explanation. I didn't get it right with ` cf-remote` obviously, thinking that I'd need to install it on every managed client. But just using it on my own machine or the CFEngine server might work for our purpose, then. I'll definitely give it a try. AFAIU it wouldn't solve the problem of regularly updating the agent, though, what a configured package repository would achieve. But I understand that there might be good reasons not to automatically update the agent.

Reasons, of course, I intend to completely ignore :laughing: .

Regarding the repo: I don't know if I'll find the time to look into it. I have a little bit of experience hosting a pacman repository and almost none of it hosting the others. But it sounds like something I could do with a little research. No promises, though.

**Reply by nickanderson** [reply]
Id just chime in that for upgrading hosts regularly, you could automate the seeding of the necessary binaries on the hub and use the standalone_self_upgrade.cf policy which is triggered when the class `default:trigger_upgrade` is defined.

That all defaults to enterprise stuff, but you can override variables and have it work with community binaries.

**Comment by basvandervlies** [comment]
To support @craigcomstock  I also use the `cf-remote` to download the community versions for example debian arm64/amd64 and then put them in our own repo server. So we are in depended of other servers and we can use `cf-remote` as fallback.  It are just tools that you can install on your stepping stone or your master policy server.

**Reply by fluchtkapsel** [reply]
I had a similar idea but I didn't want to mention it as long as I didn't know if there was some commercial interest in play. I think I'll go with that route, too.

**Comment by craigcomstock** [comment]
@MasinAD @basvandervlies thanks for the questions and comments. Sounds like this is somewhat resolved for now so I will close.

I have created https://northerntech.atlassian.net/browse/CFE-4392 to track any updates on this issue.

Feel free to watch that ticket for updates.

**Comment by atsaloli** [comment]
Automating the upgrade of the agent is very useful at scale where it's
impractical to upgrade the agent manually.

Make sure you test it first -- I find a phased approach helpful to reduce
risk of breakage (e.g, 1%, 5%, 10%, 100%).

On Thu, May 23, 2024 at 2:51 PM Nick Anderson ***@***.***>
wrote:

> Id just chime in that for upgrading hosts regularly, you could automate
> the seeding of the necessary binaries on the hub and use the
> standalone_self_upgrade.cf policy which is triggered when the class
> default:trigger_upgrade is defined.
>
> That all defaults to enterprise stuff, but you can override variables and
> have it work with community binaries.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cfengine/core/discussions/5506#discussioncomment-9540126>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD4GIG3FFWFA3YFXIKKYKLZDZQFHAVCNFSM6AAAAABHQTA72GVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TKNBQGEZDM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Achieve real learning.  Email ***@***.***
